File: apps/ios/Cursumi/Core/API/SocialAPI.swift
import Foundation

/// Chat, notas, reflexiones, reseñas, referidos, blog, materiales, juegos y
/// solicitud de instructor. Espejo de `apps/mobile/src/lib/me.ts`.
struct SocialAPI {
    let api: APIClient
    init(api: APIClient = .shared) { self.api = api }

    // MARK: Chat con el instructor

    /// Obtiene (o crea) la conversación del curso, con sus mensajes.
    func conversation(courseId: String) async throws -> Conversation {
        try await api.get("api/conversations", query: ["courseId": courseId])
    }

    func messages(conversationId: String) async throws -> [ChatMessage] {
        try await api.get("api/conversations/\(conversationId)/messages")
    }

    func sendMessage(conversationId: String, body: String) async throws -> ChatMessage {
        struct Body: Encodable { let body: String }
        return try await api.post("api/conversations/\(conversationId)/messages", Body(body: body))
    }

    func markConversationRead(_ conversationId: String) async throws {
        try await api.patch("api/conversations/\(conversationId)/read")
    }

    // MARK: Notas

    func notes(courseId: String? = nil, lessonId: String? = nil) async throws -> [Note] {
        var query: [String: String] = [:]
        if let courseId { query["courseId"] = courseId }
        if let lessonId { query["lessonId"] = lessonId }
        return try await api.get("api/notes", query: query)
    }

    func createNote(courseId: String, lessonId: String?, content: String) async throws -> Note {
        struct Body: Encodable { let courseId: String; let lessonId: String?; let content: String }
        return try await api.post("api/notes", Body(courseId: courseId, lessonId: lessonId, content: content))
    }

    func deleteNote(_ id: String) async throws {
        try await api.delete("api/notes/\(id)")
    }

    // MARK: Reflexiones y reseñas

    func reflections(courseId: String) async throws -> [Reflection] {
        struct Reply: Decodable { let reflections: [Reflection]? }
        let reply: Reply = try await api.get("api/courses/\(courseId)/learning-reflections")
        return reply.reflections ?? []
    }

    func postReflection(courseId: String, content: String) async throws {
        struct Body: Encodable { let content: String }
        try await api.post("api/courses/\(courseId)/learning-reflections", Body(content: content))
    }

    func reviews(courseId: String) async throws -> ReviewsResponse {
        try await api.get("api/courses/\(courseId)/reviews")
    }

    func postReview(courseId: String, rating: Int, comment: String?) async throws {
        struct Body: Encodable { let rating: Int; let comment: String? }
        try await api.post("api/courses/\(courseId)/reviews", Body(rating: rating, comment: comment))
    }

    // MARK: Referidos, blog, materiales

    func referral() async throws -> Referral {
        try await api.get("api/me/referral")
    }

    func blogPosts() async throws -> [BlogPostSummary] {
        let reply: BlogListResponse = try await api.get("api/blog")
        return reply.posts
    }

    func blogPost(slug: String) async throws -> BlogPost {
        try await api.get("api/blog/\(slug)")
    }

    func orgMaterials() async throws -> OrgMaterialsResponse {
        try await api.get("api/me/org-materials")
    }

    // MARK: Volverse instructor

    func applyInstructor(headline: String, bio: String, reason: String) async throws {
        struct Body: Encodable { let headline: String; let bio: String; let reason: String }
        try await api.post("api/instructor/apply", Body(headline: headline, bio: bio, reason: reason))
    }

    // MARK: Juegos en vivo (jugador)

    func joinGame(code: String, nickname: String) async throws -> String {
        struct Body: Encodable { let code: String; let nickname: String }
        struct Reply: Decodable { let gameId: String }
        let reply: Reply = try await api.post("api/games/join", Body(code: code.uppercased().trimmingCharacters(in: .whitespaces), nickname: nickname.trimmingCharacters(in: .whitespaces)))
        return reply.gameId
    }

    func game(_ gameId: String) async throws -> GameState {
        try await api.get("api/games/\(gameId)")
    }

    func answerGame(_ gameId: String, questionId: String, option: Int) async throws {
        struct Body: Encodable { let questionId: String; let selectedOption: Int }
        try await api.post("api/games/\(gameId)/answer", Body(questionId: questionId, selectedOption: option))
    }

    // MARK: Contraseña

    func changePassword(current: String, new: String) async throws {
        struct Body: Encodable { let currentPassword: String; let newPassword: String; let revokeOtherSessions: Bool }
        let res = try await api.request("POST", "api/auth/change-password", json: Body(currentPassword: current, newPassword: new, revokeOtherSessions: true))
        guard res.isOK else { throw APIError.http(res.status, message: res.errorMessage) }
    }
}

File: apps/android/app/src/main/java/com/cursumi/app/core/api/SocialApi.kt
package com.cursumi.app.core.api

import com.cursumi.app.core.model.BlogPost
import com.cursumi.app.core.model.BlogPostSummary
import com.cursumi.app.core.model.ChatMessage
import com.cursumi.app.core.model.Conversation
import com.cursumi.app.core.model.FlexibleListSerializer
import com.cursumi.app.core.model.GameState
import com.cursumi.app.core.model.Note
import com.cursumi.app.core.model.OrgMaterialsResponse
import com.cursumi.app.core.model.Referral
import com.cursumi.app.core.model.Reflection
import com.cursumi.app.core.model.ReviewsResponse
import kotlinx.serialization.Serializable

/** Chat, notas, reflexiones, reseñas, referidos, blog, materiales, juegos y solicitud de instructor. */
class SocialApi(private val api: ApiClient) {

    suspend fun conversation(courseId: String): Conversation = api.get("api/conversations", mapOf("courseId" to courseId))
    suspend fun messages(conversationId: String): List<ChatMessage> = api.get("api/conversations/$conversationId/messages")
    @Serializable private data class MsgBody(val body: String)
    suspend fun sendMessage(conversationId: String, body: String): ChatMessage = api.post("api/conversations/$conversationId/messages", MsgBody(body))
    suspend fun markConversationRead(conversationId: String) = api.patch<Unit>("api/conversations/$conversationId/read")

    suspend fun notes(courseId: String? = null, lessonId: String? = null): List<Note> {
        val q = buildMap { courseId?.let { put("courseId", it) }; lessonId?.let { put("lessonId", it) } }
        return api.get("api/notes", q)
    }
    @Serializable private data class NoteBody(val courseId: String, val lessonId: String? = null, val content: String)
    suspend fun createNote(courseId: String, lessonId: String?, content: String): Note = api.post("api/notes", NoteBody(courseId, lessonId, content))
    suspend fun deleteNote(id: String) = api.delete<Unit>("api/notes/$id")

    @Serializable private data class ReflectionsReply(val reflections: List<Reflection>? = null)
    suspend fun reflections(courseId: String): List<Reflection> = api.get<ReflectionsReply>("api/courses/$courseId/learning-reflections").reflections ?: emptyList()
    @Serializable private data class ContentBody(val content: String)
    suspend fun postReflection(courseId: String, content: String) = api.postUnit("api/courses/$courseId/learning-reflections", ContentBody(content))

    suspend fun reviews(courseId: String): ReviewsResponse = api.get("api/courses/$courseId/reviews")
    @Serializable private data class ReviewBody(val rating: Int, val comment: String? = null)
    suspend fun postReview(courseId: String, rating: Int, comment: String?) = api.postUnit("api/courses/$courseId/reviews", ReviewBody(rating, comment))

    suspend fun referral(): Referral = api.get("api/me/referral")
    suspend fun blogPosts(): List<BlogPostSummary> = api.request("GET", "api/blog").throwIfError().decode(FlexibleListSerializer(BlogPostSummary.serializer()))
    suspend fun blogPost(slug: String): BlogPost = api.get("api/blog/$slug")
    suspend fun orgMaterials(): OrgMaterialsResponse = api.get("api/me/org-materials")

    @Serializable private data class ApplyBody(val headline: String, val bio: String, val reason: String)
    suspend fun applyInstructor(headline: String, bio: String, reason: String) = api.postUnit("api/instructor/apply", ApplyBody(headline, bio, reason))

    @Serializable private data class JoinBody(val code: String, val nickname: String)
    @Serializable private data class JoinReply(val gameId: String)
    suspend fun joinGame(code: String, nickname: String): String =
        api.post<JoinBody, JoinReply>("api/games/join", JoinBody(code.uppercase().trim(), nickname.trim())).gameId
    suspend fun game(gameId: String): GameState = api.get("api/games/$gameId")
    @Serializable private data class AnswerBody(val questionId: String, val selectedOption: Int)
    suspend fun answerGame(gameId: String, questionId: String, option: Int) = api.postUnit("api/games/$gameId/answer", AnswerBody(questionId, option))
}
